p2p/shim.go

```diff
@@ -0,0 +1,330 @@
+	}
+
+	rs.BaseReactor = *NewBaseReactor(name, rs)
+
+	return rs
+}
+
+func NewChannelShim(cds *ChannelDescriptorShim, buf uint) *ChannelShim {
+	inCh := make(chan Envelope, buf)
+	outCh := make(chan Envelope, buf)
+	peerErrCh := make(chan PeerError, buf)
+
+	return &ChannelShim{
+		Descriptor: cds.Descriptor,
+		Channel:    NewChannel(ChannelID(cds.Descriptor.ID), cds.MsgType, inCh, outCh, peerErrCh),
+		InCh:       inCh,
+		OutCh:      outCh,
+		PeerErrCh:  peerErrCh,
+	}
+}
+
+// proxyPeerEnvelopes iterates over each p2p Channel and starts a separate
+// go-routine where we listen for outbound envelopes sent during Receive
+// executions (or anything else that may send on the Channel) and proxy them to
+// the corresponding Peer using the To field from the envelope.
+func (rs *ReactorShim) proxyPeerEnvelopes() {
+	for _, cs := range rs.Channels {
+		go func(cs *ChannelShim) {
+			for e := range cs.OutCh {
+				msg := proto.Clone(cs.Channel.messageType)
+				msg.Reset()
+
+				wrapper, ok := msg.(Wrapper)
+				if ok {
+					if err := wrapper.Wrap(e.Message); err != nil {
+						rs.Logger.Error(
+							"failed to proxy envelope; failed to wrap message",
+							"ch_id", cs.Descriptor.ID,
+							"msg", e.Message,
+							"err", err,
+						)
+						continue
+					}
+				} else {
+					msg = e.Message
+				}
+
+				bz, err := proto.Marshal(msg)
+				if err != nil {
+					rs.Logger.Error(
+						"failed to proxy envelope; failed to encode message",
+						"ch_id", cs.Descriptor.ID,
+						"msg", e.Message,
+						"err", err,
+					)
+					continue
+				}
+
+				switch {
+				case e.Broadcast:
+					rs.Switch.Broadcast(cs.Descriptor.ID, bz)
+
+				case !e.To.Empty():
+					src := rs.Switch.peers.Get(ID(e.To.String()))
+					if src == nil {
+						rs.Logger.Error(
+							"failed to proxy envelope; failed to find peer",
+							"ch_id", cs.Descriptor.ID,
+							"msg", e.Message,
+							"peer", e.To.String(),
+						)
+						continue
+					}
+
+					if !src.Send(cs.Descriptor.ID, bz) {
+						rs.Logger.Error(
+							"failed to proxy message to peer",
+							"ch_id", cs.Descriptor.ID,
+							"msg", e.Message,
+							"peer", e.To.String(),
+						)
+					}
+
+				default:
+					rs.Logger.Error("failed to proxy envelope; missing peer ID", "ch_id", cs.Descriptor.ID, "msg", e.Message)
+				}
+			}
+		}(cs)
+	}
+}
+
+// handlePeerErrors iterates over each p2p Channel and starts a separate go-routine
+// where we listen for peer errors. For each peer error, we find the peer from
+// the legacy p2p Switch and execute a StopPeerForError call with the corresponding
+// peer error.
+func (rs *ReactorShim) handlePeerErrors() {
+	for _, cs := range rs.Channels {
+		go func(cs *ChannelShim) {
+			for pErr := range cs.PeerErrCh {
+				if !pErr.PeerID.Empty() {
+					peer := rs.Switch.peers.Get(ID(pErr.PeerID.String()))
+					if peer == nil {
+						panic(fmt.Sprintf("failed to handle peer error; failed to find peer (%s)", pErr.PeerID))
```

Comment:
is the following situation possible: other reactor calls `StopPeerForError` on the same `Switch` and peer gets removed. however, due to congestion we are going to process `PeerErrCh` only now and find out that the peer does not exist anymore and panic! So, maybe we shouldn't panic and only log error?

Reply:
Yeah I removed the panic because we want to avoid them entirely here. RE your specific case, a reactor (now) would never call `StopPeerForError`. Only the shim (to be router) would call this.